x/auth/vesting/types/vesting_account.go

```diff
@@ -0,0 +1,619 @@
+		// exceed the desired delegation amount.
+		if coin.Amount.IsZero() || baseAmt.LT(coin.Amount) {
+			panic("delegation attempt with zero coins or insufficient funds")
+		}
+
+		// compute x and y per the specification, where:
+		// X := min(max(V - DV, 0), D)
+		// Y := D - X
+		x := sdk.MinInt(sdk.MaxInt(vestingAmt.Sub(delVestingAmt), sdk.ZeroInt()), coin.Amount)
+		y := coin.Amount.Sub(x)
+
+		if !x.IsZero() {
+			xCoin := sdk.NewCoin(coin.Denom, x)
+			bva.DelegatedVesting = bva.DelegatedVesting.Add(sdk.Coins{xCoin})
+		}
+
+		if !y.IsZero() {
+			yCoin := sdk.NewCoin(coin.Denom, y)
+			bva.DelegatedFree = bva.DelegatedFree.Add(sdk.Coins{yCoin})
+		}
+
+	}
+}
+
+// TrackUndelegation tracks an undelegation amount by setting the necessary
+// values by which delegated vesting and delegated vesting need to decrease and
+// by which amount the base coins need to increase.
+//
+// NOTE: The undelegation (bond refund) amount may exceed the delegated
+// vesting (bond) amount due to the way undelegation truncates the bond refund,
+// which can increase the validator's exchange rate (tokens/shares) slightly if
+// the undelegated tokens are non-integral.
+//
+// CONTRACT: The account's coins and undelegation coins must be sorted.
+func (bva *BaseVestingAccount) TrackUndelegation(amount sdk.Coins) {
+	for _, coin := range amount {
+		// panic if the undelegation amount is zero
+		if coin.Amount.IsZero() {
+			panic("undelegation attempt with zero coins")
+		}
+		delegatedFree := bva.DelegatedFree.AmountOf(coin.Denom)
+		delegatedVesting := bva.DelegatedVesting.AmountOf(coin.Denom)
+
+		// compute x and y per the specification, where:
+		// X := min(DF, D)
+		// Y := min(DV, D - X)
+		x := sdk.MinInt(delegatedFree, coin.Amount)
+		y := sdk.MinInt(delegatedVesting, coin.Amount.Sub(x))
+
+		if !x.IsZero() {
+			xCoin := sdk.NewCoin(coin.Denom, x)
+			bva.DelegatedFree = bva.DelegatedFree.Sub(sdk.Coins{xCoin})
+		}
+
+		if !y.IsZero() {
+			yCoin := sdk.NewCoin(coin.Denom, y)
+			bva.DelegatedVesting = bva.DelegatedVesting.Sub(sdk.Coins{yCoin})
+		}
+
+	}
+}
+
+// GetOriginalVesting returns a vesting account's original vesting amount
+func (bva BaseVestingAccount) GetOriginalVesting() sdk.Coins {
+	return bva.OriginalVesting
+}
+
+// GetDelegatedFree returns a vesting account's delegation amount that is not
+// vesting.
+func (bva BaseVestingAccount) GetDelegatedFree() sdk.Coins {
+	return bva.DelegatedFree
+}
+
+// GetDelegatedVesting returns a vesting account's delegation amount that is
+// still vesting.
+func (bva BaseVestingAccount) GetDelegatedVesting() sdk.Coins {
+	return bva.DelegatedVesting
+}
+
+// GetEndTime returns a vesting account's end time
+func (bva BaseVestingAccount) GetEndTime() int64 {
+	return bva.EndTime
+}
+
+// Validate checks for errors on the account fields
+func (bva BaseVestingAccount) Validate() error {
+	if (bva.Coins.IsZero() && !bva.OriginalVesting.IsZero()) ||
```

Comment:
Should we remove this check alltogether?

Reply:
> Should we remove this check alltogether?

Should we move this check to `NewBaseVestingAccount`? We don't want to let new accounts be initialized that fail this check.

Reply:
Yeah I think this idea is better!